text/0000-custom-dst.md

```diff
@@ -0,0 +1,279 @@
+needed (if you have both a width and a height, or if you have a width, height,
+and stride, or even further).
+
+# Detailed design
+[design]: #detailed-design
+
+The most important addition to the language shall be one new trait:
+
+```rust
+#[lang = "dst"]
+unsafe trait Dst {
+    type Meta: Copy;
+    fn size_of_val(&self) -> usize;
+}
+```
+
+The `Dst` trait should be unsafe, because it's very easy to give an incorrect
+implementation of `size_of_val`, and safe and unsafe code *must* rely on it
+being correct to not create undefined behavior.
+
+Note that it is specified that, if `size_of::<<T as Dst>::Meta>() == 0`, then
+`size_of::<*const T>()` == `size_of::<*const ()>`. This is really important for
+C compatibility, because our current compatibility with flexible arrays is
+really terrible.
+
+To actually get use of the improved fat pointers, we must add four new
+intrinsics (the names are up for bikeshedding):
+
+```rust
+// core
+mod ptr {
+    pub use intrinsics::{make_fat_ptr, make_fat_ptr_mut};
+    pub fn fat_ptr_meta<T: Dst + ?Sized>(ptr: &T) -> T::Meta {
+        unsafe {
+            core::intrinsics::fat_ptr_meta(ptr)
+        }
+    }
+}
+mod mem {
+    pub fn size_of_prelude<T: ?Sized>() -> usize {
+        unsafe {
+            core::intrinsics::size_of_prelude::<T>()
+        }
+    }
+}
+mod intrinsics {
+    extern "rust-intrinsic" {
+        fn size_of_prelude<T: ?Sized>() -> usize;
+        fn fat_ptr_meta<T: ?Sized + Dst>(ptr: *const T) -> T::Meta;
+        fn make_fat_ptr<T: Dst + ?Sized>(data: *const (),
+            meta: U::::Meta) -> *const T;
+        fn make_fat_ptr_mut<T: Dst + ?Sized>(data: *mut (),
+            meta: U::::Meta) -> *mut T;
+    }
+}
+```
+
+This is an example implementation of `[T]`:
+
+```rust
+// The last member of a struct like this must be Unsize
+#[lang = "slice"]
+struct Slice<T>([T]);
+
+unsafe impl<T> Dst for [T] {
+    type Meta = usize;
+    fn size_of_val(&self) -> usize {
+        if self.len() > 0 {
+            core::mem::size_of_val(&self[0]) * self.len()
+        } else {
+            0
+        }
+    }
+}
+
+// impl<T> Slice<T>
+impl<T> [T] {
+    pub fn len(&self) -> usize {
+        core::intrinsics::fat_ptr_meta(self)
+    }
+
+    pub fn as_ptr(&self) -> *const T {
+        unsafe {
+            &self.0 as *const [T] as *const T
+        }
+    }
+
+    pub unsafe fn from_raw_parts(buf: *const T, len: usize) -> &[T] {
+        unsafe {
+            &*core::intrinsics::make_fat_ptr(buf as *const (), len)
+        }
+    }
+}
+
+impl<T: Drop> Drop for [T] {
+    fn drop(&mut self) {
+        for el in self {
+            drop_in_place(el)
+        }
+    }
+}
+```
+
+And the following, an example implementation of a pixel buffer; this is the real
+meat of the RFC, custom fat pointers for libraries:
+
+```rust
+struct PixelBuffer([f32]);
+
+unsafe impl Dst for PixelBuffer {
+    type Meta = (usize, usize);
```

Comment:
nit: is this supposed to be a fully-general two-dimensional slice, allowing arbitrary subrectangles as subslices?  If so, don't you need a stride in a addition to the `width` and `height` in the `Meta`?